features/26-mfa/api.md

```diff
@@ -0,0 +1,975 @@
+```
+
+# Configuration Recipes
+
+## At most 1 TOTP or 1 OOB-SMS or 1 OOB-Email
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  totp:
+    maximum: 1
+  oob:
+    sms:
+      maximum: 1
+    email:
+      maximum: 1
+```
+
+## At most 1 TOTP and 1 OOB-SMS and 1 OOB-Email
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  maximum: 3
+  totp:
+    maximum: 1
+  oob:
+    sms:
+      maximum: 1
+    email:
+      maximum: 1
+```
+
+## At most 1 TOTP
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  totp:
+    maximum: 1
+```
+
+## At most 1 OOB-SMS
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  oob:
+    sms:
+      maximum: 1
+```
+
+# Endpoints
+
+## POST /mfa/totp/new
+
+### Request body
+
+```JSON
+{
+  "display_name": "my TOTP",
+  "account_name": "user@example.com",
+  "issuer": "MyApp"
+}
+```
+
+- `display_name`: The display name of the new TOTP authenticator.
+- `account_name`: If this is provided, it is used to generate `otpauth_uri`.
+- `issuer`: If this is provided, it is used to generate `otpauth_uri`.
+- `authn_session_token`: Required only when there is no access token.
+
+### Response body
+
+```JSON
+{
+  "authenticator_id": "xxx",
+  "authenticator_type": "totp",
+  "secret": "secret",
+  "otpauth_uri": "otpauth://totp/MyApp:user@example.com?secret=secret&issuer=MyApp"
+}
+```
+
+- `secret`: The secret encoded in Base32 without padding.
+- `otpauth_uri`: The [key URI](https://github.com/google/google-authenticator/wiki/Key-Uri-Format).
+
+### Description
+
+Delete all existing inactive TOTP authenticators and create a new TOTP authenticator.
+
+## POST /mfa/totp/activate
+
+### Request body
+
+```JSON
+{
+  "otp": "123456"
+}
+```
+
+- `otp`: The one-time password generated by the secret of the authenticator.
+- `authn_session_token`: Required only when there is no access token.
+
+### Response body
+
+```JSON
+{
+  "recovery_codes": ["xxx", "xxx"]
+}
+```
+
+- `recovery_codes`: Present only when the activated authenticator is the first active authenticator.
+
+### Description
+
+Activate the just created inactive TOTP authenticator.
+
+## GET /mfa/totp/qrcode
+
+## Query parameters
+
+- `otpauth_uri`: The [key URI](https://github.com/google/google-authenticator/wiki/Key-Uri-Format).
+- `account_name`: The [account name](https://github.com/google/google-authenticator/wiki/Key-Uri-Format#label).
+- `secret`: The [Base32 encoded secret](https://github.com/google/google-authenticator/wiki/Key-Uri-Format#secret).
+- `issuer`: The [issuer](https://github.com/google/google-authenticator/wiki/Key-Uri-Format#label).
+
+If `otpauth_uri` is given, it is used as is. Otherwise `otpauth_uri` is generated from `account_name`, `secret` and `issuer`.
+
+This endpoint accepts `secret` instead of `authenticator_id` because this endpoint is supposed to be used directly,
+such as setting it to the `src` of `<img>`. So it is stateless and does not require authentication.
```

Comment:
Oops, I totally didn't get it before you spell it out.

Ummm, do we really want to encourage that? Coz I think embed the secret in image URL as GET Parameter is not a good practices and we shouldn't encourage users to do so... (since URL could be tracked by browsers / extensions / server side logs etc)

I think it is better to display the QR code by:

1. Users call the APIs, get some sort of data:image/png format, and render it, OR
1. Get the png from a stateful image URL

@louischan-oursky or you have done more research and found that it is the best practices?

Reply:
> Users call the APIs, get some sort of data:image/png format, and render it

Here is my proposal

-  Remove `/mfa/totp/qrcode` endpoint so there is no chance the secret appear in URL. 
-  The QR code is a image representation of the `otpauth_uri`. `otpauth_uri` includes `secret` in its query string. The three are equivalent. Knowing one of them allows deriving the secret. We want to make the API both flexible and ergonomic. Therefore, in `/mfa/totp/new`, we should 
    - Include `secret`, which allows the developer to display it in a way they want.
    - Include `otpauth_uri`, which is correctly formatted Key URI so that it is ready to generate QR code in the developer's own way.
    - Include `qr_code_image_uri`, which is data URI of a server generated QR code image of the secret. In most case, the developer just needs this.

None of the above is optional. They always appear in the response.

> Get the png from a stateful image URL

The secret should not be ever revealed again. Its sole appearance is in the response of `/mfa/totp/new`